### PR TITLE
Ofrich/add context change hooks

### PR DESCRIPTION
**Pull request by ofrichen**
### **User description**
Added `on_context_change` hook, to be able to know when a tenant changes and report metrics/throw an error when needed.

___

### **PR Type**
Enhancement

___

### **Description**
- Added `on_context_change` hook for tenant change notifications

- Implemented callback management methods for subscription/unsubscription

- Added comprehensive test coverage for context change scenarios

- Enabled monitoring of both tenant and cross-tenant flag changes

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["Tenant Change"] --> B["Context State Check"]
  B --> C["Notify Callbacks"]
  C --> D["Execute Registered Hooks"]
  E["Cross-Tenant Flag Change"] --> B
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>multitenant.rb</strong><dd><code>Context change hook implementation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

lib/multitenant.rb

<ul><li>Added <code>CONTEXT_CHANGE_CALLBACKS</code> array for storing callback functions<br> <li> Modified <code>current_tenant=</code> and <code>allow_dangerous_cross_tenants=</code> to trigger <br>notifications<br> <li> Implemented <code>on_context_change</code>, <code>remove_on_context_change</code>, and <br><code>clear_on_context_change</code> methods<br> <li> Added <code>current_context_state</code> and <code>notify_context_change</code> helper methods</ul>

</details>

  </td>
  <td><a href="https://github.com/DaPulse/multitenant/pull/16/files#diff-d73fc74a125d8e67bd08f2d50a9aa64b79bc683a775c9925c11efe483cd9cb06">+35/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>multitenant_spec.rb</strong><dd><code>Context change callback test coverage</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

spec/multitenant_spec.rb

<ul><li>Added comprehensive test suite for context change callbacks<br> <li> Tests cover tenant changes, cross-tenant flag changes, and callback <br>management<br> <li> Includes edge cases like same-value assignments and multiple callbacks<br> <li> Tests integration with existing methods like <code>with_tenant</code> and <br><code>dangerous_cross_tenants</code></ul>

</details>

  </td>
  <td><a href="https://github.com/DaPulse/multitenant/pull/16/files#diff-30a8c67f6502e25cd627ee4996da64a5f88eb82130d63c3094ad88ca9f0ec42e">+192/-0</a>&nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

### CODE REVIEW

File: lib/multitenant.rb [hunk 1]
```diff
@@ -43,18 +63,42 @@ def with_tenant(tenant, options = {}, &block)
...
+
+    def current_context_state
+      return {
+        tenant: Thread.current[CURRENT_TENANT],
+        is_cross_tenant: (Thread.current[ALLOW_DANGEROUS] == true)
+      }
+    end
+
+    def notify_context_change(previous_state, current_state)
+      return if Thread.current[:multitenant_suppress_ctx_change].to_i > 0 || previous_state == current_state
```
Comment (omerdoron1) [inline lib/multitenant.rb]: split into two " if return " lines

File: lib/multitenant.rb [hunk 2]
```diff
@@ -43,18 +63,42 @@ def with_tenant(tenant, options = {}, &block)
       Multitenant.extra_tenant_ids = options[:extra_tenant_ids] if options[:extra_tenant_ids]
       yield
     ensure
-      Multitenant.current_tenant = previous_tenant
+      without_context_change_notifications do
```
Comment (omerdoron1) [inline lib/multitenant.rb:71]: add comment: prevent notifications on scope return to the original scope

File: lib/multitenant.rb [hunk 3]
```diff
@@ -43,18 +63,42 @@ def with_tenant(tenant, options = {}, &block)
...
+
     def dangerous_cross_tenants(&block)
       previous_value = Multitenant.allow_dangerous_cross_tenants
       Multitenant.allow_dangerous_cross_tenants = true
       Multitenant.with_tenant(nil) do
         yield
       end
     ensure
-      Multitenant.allow_dangerous_cross_tenants = previous_value
+      without_context_change_notifications do
```
Comment (omerdoron1) [inline lib/multitenant.rb:85]: same comment

File: lib/multitenant.rb [hunk 4]
```diff
@@ -9,21 +9,28 @@ class << self
...
 
     def allow_dangerous_cross_tenants
       Thread.current[ALLOW_DANGEROUS]
     end
 
     def allow_dangerous_cross_tenants=(value)
+      previous_state = current_context_state
       Thread.current[ALLOW_DANGEROUS] = value
+      current_state = current_context_state
+      notify_context_change(previous_state, current_state)
```
Comment (omerdoron1) [inline lib/multitenant.rb:35]: return original value

File: lib/multitenant.rb [hunk 5]
```diff
@@ -9,21 +9,28 @@ class << self
...
 
     def current_tenant
       Thread.current[CURRENT_TENANT]
     end
 
     def current_tenant=(value)
+      previous_state = current_context_state
       Thread.current[CURRENT_TENANT] = value
+      current_state = current_context_state
+      notify_context_change(previous_state, current_state)
```
Comment (omerdoron1) [inline lib/multitenant.rb:22]: return original value

File: lib/multitenant.rb [hunk 6]
```diff
@@ -9,21 +9,28 @@ class << self
     CURRENT_TENANT = 'Multitenant.current_tenant'.freeze
     ALLOW_DANGEROUS = 'Multitenant.allow_dangerous_cross_tenants'.freeze
     EXTRA_TENANT_IDS = 'Multitenant.extra_tenant_ids'.freeze
+    CONTEXT_CHANGE_CALLBACKS = []
```
Comment (omerdoron1) [inline lib/multitenant.rb:12]: Multitenant.add_on_context_change_hook do |prev, curr|
    MultitenantChangeHandler.handle(prev, curr)
end

**Review by omerdoron1 — APPROVED** [review]

File: lib/multitenant.rb [hunk 7]
```diff
@@ -54,7 +81,32 @@ def dangerous_cross_tenants(&block)
...
     ensure
-      Multitenant.allow_dangerous_cross_tenants = previous_value
+      # Suppress context change notifications when restoring the previous tenant
+      without_context_change_notifications do
+        Multitenant.allow_dangerous_cross_tenants = previous_value
+      end
+    end
+
+    def without_context_change_notifications
+      Thread.current[:multitenant_suppress_ctx_change] = Thread.current[:multitenant_suppress_ctx_change].to_i + 1
```
Comment (erez-ra) [inline lib/multitenant.rb:91]: just to make sure I understand
```
Thread.current[:multitenant_suppress_ctx_change]
```
is not initilized to 0 anywhere right? we rely on `nil.to_i == 0` ?

**Review by erez-ra — APPROVED** [review]

File: lib/multitenant.rb (hunk 7 above)
Comment (ofrichen) [reply lib/multitenant.rb:91]: correct
